include/taskolib/exceptions.h

```diff
@@ -48,57 +49,44 @@ namespace task {
  * {
  *     std::cerr << e.what() << "\n";
  * }
- * \endcode
  *
- * \note
- * task::Error is derived from std::runtime_error. It can therefore be caught by
- * `catch (const std::exception&)`, `catch (const std::runtime_error&)`, and
- * `catch (const task::Error&)`.
- */
-class Error : public std::runtime_error
-{
-public:
-    using std::runtime_error::runtime_error;
-};
-
-/**
- * An exception class storing the index of the step in which an error occurred, in
- * addition to the error message.
- *
- * \code
  * try
  * {
- *     throw task::ErrorAtIndex("An error has occurred", 42);
+ *     throw task::Error("An error has occurred", 42);
  * }
- * catch (const task::ErrorAtIndex& e)
+ * catch (const task::Error& e)
  * {
- *     std::cerr << e.what() << " in step " << e.get_index() << "\n";
+ *     std::cerr << e.what();
+ *     auto maybe_step_index = e.get_index();
+ *     if (maybe_step_index)
+ *         std::cerr << ": step index " << *maybe_step_index;
+ *     std::cerr << "\n";
  * }
  * \endcode
  *
  * \note
- * task::ErrorAtIndex inherits from std::runtime_error. It can therefore be caught by
- * `catch (const std::exception&)`, `catch (const std::runtime_error&)`,
- * `catch (const task::Error&)`, and `catch (const task::ErrorAtIndex&)`.
+ * task::Error is derived from std::runtime_error. It can therefore be caught by
+ * `catch (const std::exception&)`, `catch (const std::runtime_error&)`, and
+ * `catch (const task::Error&)`.
  */
-class ErrorAtIndex : public Error
+class Error : public std::runtime_error
 {
 public:
-    ErrorAtIndex(const std::string& msg, StepIndex index)
-        : Error(msg)
-        , index_(index)
+    explicit Error(const std::string& msg, OptionalStepIndex opt_step_index = gul14::nullopt)
```

Comment:
I thought I understood `explicit`, but...
even with the following patch applied the tests run through just fine, despite the fact that I changed the test like so:

![image](https://user-images.githubusercontent.com/16012374/207031223-a986a22b-9707-4e75-9c61-04b646c6d7c6.png)

Where the 1st arg is not a `string` and the second is no `int` :thinking: 

```diff
diff --git a/include/taskolib/exceptions.h b/include/taskolib/exceptions.h
index 97c60c5..1d599af 100644
--- a/include/taskolib/exceptions.h
+++ b/include/taskolib/exceptions.h
@@ -77,11 +77,6 @@ public:
         , index_{ opt_step_index }
     {}
 
-    explicit Error(const char* msg, OptionalStepIndex opt_step_index = gul14::nullopt)
-        : std::runtime_error(msg)
-        , index_{ opt_step_index }
-    {}
-
     /// Return the associated step index.
     OptionalStepIndex get_index() const { return index_; }
 
```

Reply:
Ah, the `explicit` prohibits

```diff
-    const task::Error e{ "Test" };
+    const task::Error e = "Test";
```

which I believe should be forbidden.

Reply:
Got it. And with the optional `OptionalStepIndex` it must/should be `explicit`.

Reply:
BTW why do you declare `double` as `StepIndex` in the snippet above?

```c++
TEST_CASE("Error: Copy assignment...)
{
    const task::Error e("Test", 1.0);
...
}
```

To test an implicit conversion to `uint16_t`? And the compiler didn't complain?

Reply:
Right. Well, `explicit` just dis-allows implicit conversion, but here
```c++
auto x = task::Error{ a, b };
```
the creation of the Error object is explicit. It does not mean implicit conversion of the individual constructor parameters.

```c++
float f = 3; // implicit conversion from `int` to `float`
Error e = "convert"; // implicit conversion of `char const*` to `Error`

auto e = Error{ "convert", 3.0 }; // explicitly construct an Error from a C string and a float
```

Reply:
Ok. Thx!

Reply:
Just to add a link to the C++ Core Guidelines: [C.46: By default, declare single-argument constructors explicit](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-explicit)

(Our constructor is *practically* a single-argument constructor because of the default argument.)